src/lib.rs

```diff
@@ -76,78 +77,166 @@ extern crate enumflags2;
 };
 pub use fs::{path_beneath_rules, AccessFs, PathBeneath, PathFd};
-pub use ruleset::{Access, RestrictionStatus, Rule, Ruleset, RulesetCreated, RulesetStatus};
-use ruleset::{PrivateAccess, PrivateRule};
+pub use ruleset::{RestrictionStatus, Rule, Ruleset, RulesetCreated, RulesetStatus};
 
+#[cfg(test)]
+pub use compat::{can_emulate, landlock_is_enabled};
 #[cfg(test)]
 pub use errors::TestRulesetError;
 
+use access::PrivateAccess;
+use compat::{CompatState, Compatibility, TryCompat};
+use ruleset::PrivateRule;
+
+#[cfg(test)]
+use strum::IntoEnumIterator;
+
+mod access;
 mod compat;
 mod errors;
 mod fs;
 mod ruleset;
 mod uapi;
 
-#[cfg(all(test, not(feature = "test-without-kernel-support")))]
+#[cfg(test)]
 mod tests {
     use crate::*;
 
-    fn ruleset_root_compat() -> Result<RestrictionStatus, TestRulesetError> {
-        Ok(Ruleset::new()
-            .handle_access(AccessFs::from_all(ABI::V1))?
-            .create()?
-            .add_rule(PathBeneath::new(
-                PathFd::new("/")?,
-                AccessFs::from_all(ABI::V1),
-            ))?
-            .restrict_self()?)
-    }
+    // Emulate old kernel supports.
+    fn check_ruleset_support<F>(partial: ABI, full: ABI, check: F, error_if_abi_lt_partial: bool)
+    where
+        F: Fn(Ruleset) -> Result<RestrictionStatus, TestRulesetError>,
+    {
+        // If there is no partial support, it means that `full == partial`.
+        assert!(partial <= full);
+        for abi in ABI::iter() {
+            let ret = check(Ruleset::from(abi));
 
-    fn ruleset_root_fragile() -> Result<RestrictionStatus, TestRulesetError> {
-        // Sets default support requirement: abort the whole sandboxing for any Landlock error.
-        Ok(Ruleset::new()
-            // Must have at least the execute check…
-            .set_best_effort(false)
-            .handle_access(AccessFs::Execute)?
-            // …and possibly others.
-            .set_best_effort(true)
-            .handle_access(AccessFs::from_all(ABI::V1))?
-            .create()?
-            .set_no_new_privs(true)
-            .add_rule(PathBeneath::new(
-                PathFd::new("/")?,
-                AccessFs::from_all(ABI::V1),
-            ))?
-            .restrict_self()?)
+            // Useful for failed tests and with cargo test -- --show-output
+            println!("Checking ABI {:?}: {:#?}", abi, ret);
+            if can_emulate(abi, full) {
+                if abi < partial && error_if_abi_lt_partial {
+                    // TODO: Check exact error type; this may require better error types.
+                    assert!(matches!(ret, Err(TestRulesetError::Ruleset(_))));
+                } else {
+                    assert!(matches!(
+                        ret,
+                        Ok(RestrictionStatus {
+                            ruleset: ruleset_status,
+                            no_new_privs: true
+                        }) if ruleset_status == if abi >= full {
```

Comment:
Maybe it's just me, but I find nesting this if inside the comparison inside a match to be pretty annoying to read and reason about.  I think the code would be more readable if the whole "if abi >= full" clause with its branches was pulled out before the assert, to keep the arguments put into the nested macro calls more simple and straightforward.

Reply:
Indeed, I'll move this one and the other `if … == if …` out of the `matches!()` call.